Unified PDF: Find highlights for PDF content with fills obscure the yellow background

hortont424 opened the pull request:
#### 126a5c9aa7aa6dca80eb9d9f00ba98c97ac34732
<pre>
Unified PDF: Find highlights for PDF content with fills obscure the yellow background
<a href="https://bugs.webkit.org/show_bug.cgi?id=270287">https://bugs.webkit.org/show_bug.cgi?id=270287</a>
<a href="https://rdar.apple.com/123758797">rdar://123758797</a>

Reviewed by Sammy Gill.

Because PDFKit doesn&apos;t have a concept of &quot;range only painting&quot; like WebKit,
we can&apos;t actually depend on knocking out the background to paint the TextIndicator.

As such, we instead do what PDFKit does: put the backgrounds back, and multiply the yellow on top!

Also, fix the snapshot coordinate space, which was broken by a non-find-related
earlier patch.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::UnifiedPDFPlugin::textIndicatorForSelection):

Canonical link: <a href="https://commits.webkit.org/275501@main">https://commits.webkit.org/275501@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3b77efe6574c437408c7dff969e0b934958a58ca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/41981 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/20999 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/44375 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/44567 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/38092 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/24168 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/18327 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/34785 "Failure limit exceed. At least found 164 new test failures: accessibility/math-multiscript-attributes.html, animations/additive-transform-animations.html, animations/multiple-backgrounds.html, compositing/canvas/accelerated-canvas-compositing-size-limit.html, compositing/tiling/sticky-change-to-tiled.html, css1/formatting_model/inline_elements.html, css2.1/20110323/absolute-non-replaced-height-002.htm, css2.1/20110323/absolute-non-replaced-height-007.htm, css2.1/20110323/absolute-non-replaced-height-009.htm, css2.1/20110323/absolute-non-replaced-max-height-002.htm ... (failure)") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/42555 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/17925 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/36151 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/15718 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/15599 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/45982 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/38169 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/37509 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/41419 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/16797 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/13804 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/39945 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/18416 "Built successfully") | | | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/9414 "The change is no longer eligible for processing.") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/18476 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5639 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/18061 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

sammygill commented on Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
```diff
@@ -2823,10 +2822,23 @@ static NSStringCompareOptions compareOptionsForFindOptions(WebCore::FindOptions
+#if PLATFORM(MAC)
+    Color highlightColor = *roundAndClampToSRGBALossy([NSColor findHighlightColor].CGColor);
+#else
+    Color highlightColor = SRGBA<float>(.99, .89, 0.22, 1.0);
```
It's a little hard to figure out what this `SRGBA` object is supposed to represent as opposed to the one above where the names make it easier to see what it's supposed to be the color for the find highlight. Could we somehow make this one a little clearer?

hortont424 committed:
Unified PDF: Find highlights for PDF content with fills obscure the yellow background
https://bugs.webkit.org/show_bug.cgi?id=270287
rdar://123758797

Reviewed by Sammy Gill.

Because PDFKit doesn't have a concept of "range only painting" like WebKit,
we can't actually depend on knocking out the background to paint the TextIndicator.

As such, we instead do what PDFKit does: put the backgrounds back, and multiply the yellow on top!

Also, fix the snapshot coordinate space, which was broken by a non-find-related
earlier patch.

* Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm:
(WebKit::UnifiedPDFPlugin::paintPDFContent):
(WebKit::UnifiedPDFPlugin::textIndicatorForSelection):

Canonical link: https://commits.webkit.org/275501@main
